plugins/field-grid-dropdown/src/index.ts

```diff
@@ -9,169 +9,215 @@
+   *   or a function which generates these options.
+   * @param validator A function that is called to validate
+   *  changes to the field's value. Takes in a language-neutral dropdown
+   *  option & returns a validated language-neutral dropdown option, or null
+   *  to abort the change.
+   * @param config A map of options used to configure the field.
+   *  See the [field creation documentation]{@link
+   * https://developers.google.com/blockly/guides/create-custom-blocks/fields/built-in-fields/dropdown#creation}
+   *  for a list of properties this parameter supports.
+   * @extends {Blockly.Field}
+   * @constructor
    * @throws {TypeError} If `menuGenerator` options are incorrectly structured.
    */
-  constructor(menuGenerator, validator = undefined, config = undefined) {
+  constructor(menuGenerator: Blockly.MenuGenerator,
+      validator?: FieldGridDropdownValidator,
+      config?: FieldGridDropdownConfig) {
     super(menuGenerator, validator, config);
 
-    /**
-     * The number of columns in the dropdown grid. Must be an integer value
-     * greater than 0. Defaults to 3.
-     * @type {number}
-     * @private
-     */
-    this.columns_ = 3;
-    if (config && config['columns']) {
-      this.setColumnsInternal_(config['columns']);
+    if (config?.columns) {
+      this.setColumnsInternal_(config.columns);
     }
 
-    if (config && config['primaryColour']) {
-      this.primaryColour = config['primaryColour'];
+    if (config && config.primaryColour) {
+      this.primaryColour = config.primaryColour;
     }
 
-    if (config && config['borderColour']) {
-      this.borderColour = config['borderColour'];
+    if (config && config.borderColour) {
+      this.borderColour = config.borderColour;
     }
   }
 
   /**
    * Constructs a FieldGridDropdown from a JSON arg object.
-   * @param {!Object} options A JSON object with options.
-   * @returns {!FieldGridDropdown} The new field instance.
+   *
+   * @param config A JSON object with options.
+   * @returns The new field instance.
    * @package
    * @nocollapse
    */
-  static fromJson(options) {
-    // `this` might be a subclass of FieldGridDropdown if that class doesn't
+  static fromJson(config: FieldGridDropdownFromJsonConfig) {
+    if (!config.options) {
+      throw new Error(
+          'options are required for the dropdown field. The ' +
+          'options property must be assigned an array of ' +
+          '[humanReadableValue, languageNeutralValue] tuples.');
+    }
+    // `this` might be a subclass of FieldDropdown if that class doesn't
     // override the static fromJson method.
-    return new this(options['options'], undefined, options);
-  }
-
-  /**
-   * Sets the number of columns on the grid. Updates the styling to reflect.
-   * @param {number} columns The number of columns. Is rounded to
-   *    an integer value and must be greater than 0. Invalid
-   *    values are ignored.
-   * @private
-   */
-  setColumns(columns) {
```

Comment:
Why was this removed?

Reply:
I'll need to check back on this later this week when I've got time. Some of the changes were pulled from previous work I'd done before the `9.3.1` updates, so if this plugin changed in the meantime it's possible I missed something.

Reply:
I see now - it was marked `private`, though it's not used anywhere internally. There's no purpose in it being here if it's private as it's currently not used and can't be used by subclasses.

I could add it back as public, or I could leave it as is. WDYT?

Reply:
If it works (i.e. properly rerenders the dropdown div) I think it makes sense to readd it as public! Seems like a reasonable part of the API this field should have =)

But if it doesn't actually work properly (and can't trivially be fixed). Then I'd say leave as is (i.e. removed). That SGTY?

Reply:
SGTM! It works just fine as public, so I'll add it back.